### ISSUE TITLE
Feature: Decorator-style Commander

### ISSUE THREAD

**Issue by darrenrahnemoon** [tj/commander.js#1596]
So for a personal project I developed a form of syntactic sugar for writing and structuring commands for commander.js. 
It's typescript based and uses decorators and `reflect-metadata` to store and generate commander `Command` instances. 

An example of what one of my existing commands look like: 

```
@CommandGroup('services')
export class DockerServices {

	@Command()
	@Argument('<group>', 'name of the overlaying group')
	@Argument('[services...]', 'name of the individual services')
	@Option('--environment, -e', 'environment', {
		choices  : [ 'development', 'testing', 'production' ],
		defaultValue : 'development',
	})
	async start(services: string[] = [], { environment = 'development' } = {}) {
		// command action
	}

}

new DockerServices().toCommand();  // gives out a command that you can attach to your program
```

Was wondering what everyone thinks about this format? 
Is it worth creating a spin-off package (like `commander-decorator` or `commander-class`) or should I try to get it integrated into this package and put up a PR?

**Comment by shadowspawn** [comment]
> should I try to get it integrated into this package and put up a PR?

Welcome to discuss here, but after a little research it does not look like we would consider adding decorator support to Commander anytime soon. Only at stage 2 with TC39, experimental in TypeScript, and no support yet in Node.js.

- https://github.com/tc39/proposal-decorators
- https://www.typescriptlang.org/docs/handbook/decorators.html
- https://node.green/

-----

https://google.github.io/styleguide/tsguide.html#decorators

> Do not define new decorators

>Why?
> We generally want to avoid decorators, because they were an experimental feature that have since diverged from the TC39 proposal and have known bugs that won't be fixed.

-----

https://medium.com/@martin_hotell/10-typescript-pro-tips-patterns-with-or-without-react-5799488d6680

> 6. Don’t use decorators for class Components

> TypeScript uses old version of decorator proposal which isn’t gonna be implemented in ECMAscript standard

> What is most important though, in terms of type checking within JSX, is, that decorators don’t extend class type definition.

-----

https://github.com/microsoft/TypeScript/issues/4881#issuecomment-858435857

> tl;dr - this feature is jammed up in TC39. Can't blame the TS people for this at all. They're not gonna implement until it's standard.

https://github.com/microsoft/TypeScript/issues/4881#issuecomment-608084490

>  I'm certain that the TypeScript team won't implement or consider anything in this regard until the decorator proposal is finished and supported by the browsers.

**Comment by shadowspawn** [comment]
Related: https://github.com/mattstrom/commander-ts

**Comment by jingyuexing** [comment]
> 相关：[https](https://github.com/mattstrom/commander-ts) : [//github.com/mattstrom/commander-ts](https://github.com/mattstrom/commander-ts)

There is always someone in this world who will keep up with your thoughts :)

**Comment by jmcdo29** [comment]
Sorry for necro-posting on this issue, but I created [nest-commander](https://nest-commander.jaymcdoniel.dev) as a commander wrapper for the [NestJS framework](https://nestjs.com). Decorator support and Nest's DI all wrapped into a quick integration package
